homeassistant/components/arcticspa/config_flow.py

```diff
@@ -0,0 +1,97 @@
+"""Config flow for Arctic Spa integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from pyarcticspas import Spa
+from pyarcticspas.error import SpaHTTPException, TooManyRequestsError, UnauthorizedError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_API_KEY
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): str,
+    }
+)
+
+
+def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
```

Comment:
This function is fairly small. I think we can move this into the config flow step completely; this allows you to get rid of the custom exceptions.

Reply:
I'll have to think about this one a bit as I'm not well-versed in Python testing.

Currently, `validate_input` is patched during testing so I can test for things like invalid tokens or successful validation using mock API keys.

If I move the function into `async_step_user` (which is the only place where it's used), I'll have to mock the whole `asyn_step_user` function.

Alternatively, if the main reason to fold it into `async_step_user` is because it's a small function, I'd rather keep it around so I can do more granular test cases.

Reply:
You should patch your library :)

Reply:
D'oh. As I said, I'm fairly new to Python tests.

Library patched, `validate_input` removed.